strapi/strapi
# fix(graphql): inherit publication state for i18n localizations

### What does it do?

- Updated GraphQL relation resolution so nested localizations inherit the publication state requested by the root built-in query.
- This keeps default/live GraphQL queries from leaking draft localizations while preserving draft/preview localization queries.
- Added regression coverage for both live and draft GraphQL localization queries.

### Why is it needed?

- There is a discrepancy between localization results returned by **REST** API and **GraphQL** API.  
- GraphQL is returning unpublished/previous revisions of the entity.

#### REST
Endpoint: http://localhost:1337/api/homepage?populate=localizations

SQL: 
```sql
select `t0`.*, `t0`.`document_id` from `homepages` as `t0` where (`t0`.`document_id` in ('puxnj5x7yplx0qig9wu40kom') and `t0`.`id` not in (2) and `t0`.`published_at` is not null);
```
Result:
```json
{
    "data": {
        "id": 2,
        "documentId": "puxnj5x7yplx0qig9wu40kom",
        "title": "home",
        "slug": "homepage",
        "createdAt": "2024-11-01T00:48:17.869Z",
        "updatedAt": "2024-11-01T00:48:33.244Z",
        "publishedAt": "2024-11-01T00:48:17.876Z",
        "locale": "en",
        "localizations": [
            {
                "id": 4,
                "documentId": "puxnj5x7yplx0qig9wu40kom",
                "title": "inicio",
                "slug": "homepage",
                "createdAt": "2024-11-01T00:48:33.233Z",
                "updatedAt": "2024-11-01T00:48:33.233Z",
                "publishedAt": "2024-11-01T00:48:33.240Z",
                "locale": "es"
            }
        ]
    },
    "meta": {}
}
```

#### GraphQL
Endpoint: http://localhost:1337/graphql

SQL: 
```sql
select `t0`.*, `t0`.`document_id` from `homepages` as `t0` where (`t0`.`document_id` in ('puxnj5x7yplx0qig9wu40kom') and `t0`.`id` not in (2)) limit 25;
```
GraphQL Query: 
```graphql
query Homepage {
    homepage {
        title
        slug
        locale
        localizations {
            locale
            title
            slug
            publishedAt
        }
    }
}
```
Result:
```json
{
    "data": {
        "homepage": {
            "title": "home",
            "slug": "homepage",
            "locale": "en",
            "localizations": [
                {
                    "locale": "en",
                    "title": "home",
                    "slug": "homepage",
                    "publishedAt": null
                },
                {
                    "locale": "es",
                    "title": "inicio",
                    "slug": "homepage",
                    "publishedAt": null
                },
                {
                    "locale": "es",
                    "title": "inicio",
                    "slug": "homepage",
                    "publishedAt": "2024-11-01T00:48:33.240Z"
                }
            ]
        }
    }
}
```

### How to test it?

1. Add i18n content to CMS, for example: Homepage
2. Make a GraphQL query for homepage and its localizations:
```
query Homepages {
    homepage {
        title
        slug
        locale
        localizations {
            locale
            title
            slug
            publishedAt
        }
    }
}
```
3. Receive expected result, with only published localizations:
```json
{
    "data": {
        "homepage": {
            "title": "home",
            "slug": "homepage",
            "locale": "en",
            "localizations": [
                {
                    "locale": "es",
                    "title": "inicio",
                    "slug": "homepage",
                    "publishedAt": "2024-11-01T00:48:33.240Z"
                }
            ]
        }
    }
}
```

The underlying SQL command will be:
```sql
select `t0`.*, `t0`.`document_id` from `homepages` as `t0` where (`t0`.`document_id` in ('puxnj5x7yplx0qig9wu40kom') and `t0`.`id` not in (2) and `t0`.`published_at` is not null) limit 25;
```

### Related issue

#21273
